lib/style-engine/class-wp-style-engine-gutenberg.php

```diff
@@ -0,0 +1,157 @@
+
+	/**
+	 * Style definitions that contain the instructions to
+	 * parse/output valid Gutenberg styles from a block's attributes.
+	 * For every style definition, the follow properties are valid:
+	 *
+	 *  - property_key => the key that represents a valid CSS property, e.g., "margin" or "border".
+	 *  - value_func   => a function to generate an array of valid CSS rules for a particular style object.
+	 *                    For example, `'padding' => 'array( 'top' => '1em' )` will return `array( 'padding-top' => '1em' )`
+	 */
+	const BLOCK_STYLE_DEFINITIONS_METADATA = array(
+		'spacing' => array(
+			'padding' => array(
+				'property_key' => 'padding',
+				'value_func'   => 'gutenberg_get_style_engine_css_box_rules',
+			),
+			'margin'  => array(
+				'property_key' => 'margin',
+				'value_func'   => 'gutenberg_get_style_engine_css_box_rules',
+			),
+		),
+	);
+
+	/**
+	 * Utility method to retrieve the main instance of the class.
+	 *
+	 * The instance will be created if it does not exist yet.
+	 *
+	 * @return WP_Style_Engine_Gutenberg The main instance.
+	 */
+	public static function get_instance() {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Returns a CSS ruleset based on the instructions in BLOCK_STYLE_DEFINITIONS_METADATA.
+	 *
+	 * @param string|array  $style_value A single raw Gutenberg style attributes value for a CSS property.
+	 * @param array<string> $path        An array of strings representing a path to the style value.
+	 *
+	 * @return array A CSS ruleset compatible with add_style().
+	 */
+	protected function get_block_style_css_rules( $style_value, $path ) {
+		$style_definition = _wp_array_get( static::BLOCK_STYLE_DEFINITIONS_METADATA, $path, null );
+
+		if ( ! empty( $style_definition ) ) {
+			if (
+				isset( $style_definition['value_func'] ) &&
+				is_callable( $style_definition['value_func'] )
+			) {
+				return call_user_func( $style_definition['value_func'], $style_value, $style_definition['property_key'] );
+			}
+		}
+
+		return array();
+	}
+
+	/**
+	 * Returns an CSS ruleset destined to be inserted in an HTML `style` attribute.
+	 * Styles are bundled based on the instructions in BLOCK_STYLE_DEFINITIONS_METADATA.
+	 *
+	 * @param array $block_styles An array of styles from a block's attributes.
+	 * @param array $options = array(
+	 *     'inline' => (boolean) Whether to return inline CSS rules.
+	 * );.
+	 *
+	 * @return string A CSS ruleset formatted to be placed in an HTML `style` attribute.
+	 */
+	public function generate( $block_styles, $options = array() ) {
+		$output = '';
+
+		if ( empty( $block_styles ) ) {
+			return $output;
+		}
+
+		$rules = array();
+
+		foreach ( $block_styles as $block_style => $block_style_properties ) {
+			if ( empty( $block_style ) || ! is_array( $block_style_properties ) ) {
+				continue;
+			}
+			$path        = array( $block_style, key( $block_style_properties ) );
```

Comment:
`$block_style_properties` is an array, so `key` here appears to be returning `margin` for me even if the `$block_style_properties` contains both `margin` and `padding` values.

Instead, do we need to loop over `$block_style_properties` in another for loop?

Reply:
Related to my complete brain short-circuit this morning. 🙇